### PR TITLE
Check gnome version before graphical reboot

### PR DESCRIPTION
**Pull request by ggkioulis**
GNOME 3.38 changes the openQA graphical reboot process.
This requires branching the reboot behavior based on gnome version and not  on product version, since the afforementioned behavior can affect the maintenance products eventually.

- Related ticket: https://progress.opensuse.org/issues/77785
- Needles: No Needles
- Verification run: [sle-15-SP2](http://angmar.suse.de/tests/2629#step/reboot_gnome/11) |  [sle-12-SP4](http://angmar.suse.de/tests/2628#step/reboot_gnome/11) | [upgrade-Leap-to-Tumbleweed](http://angmar.suse.de/tests/2630#step/reboot_gnome/9) | [Tumbleweed](http://angmar.suse.de/tests/2627#step/reboot_and_install/12) | [Tumbleweed-Staging](http://angmar.suse.de/tests/2626#step/reboot_gnome/10)

### CODE REVIEW

File: lib/power_action_utils.pm
Comment (DimStar77) [inline lib/power_action_utils.pm:86]: Why elevate to root? a regular user can very well read the rpm database

File: lib/power_action_utils.pm
Comment (DimStar77) [inline lib/power_action_utils.pm:91]: The change happened with gnome 3.38.0 (well, actually 3.37.x, but if you go 3.38.0 it's probably closest to reality/relevance to products)

File: lib/power_action_utils.pm
Comment (ggkioulis) [reply lib/power_action_utils.pm:86]: It's not to read the rpm database, I had this [issue](http://angmar.suse.de/tests/2625#step/reboot_gnome/4)

File: lib/power_action_utils.pm
Comment (ggkioulis) [reply lib/power_action_utils.pm:91]: Thanks!

File: lib/power_action_utils.pm
Comment (DimStar77) [reply lib/power_action_utils.pm:86]: oh.. that's … unexpected; as we write to /dev/ttyS0 all the time.

And this likely only happens in your test case as you shortcut too much. you skip the consoletest_setup, that assigns the proper permissions

Comment (ggkioulis) [reply lib/power_action_utils.pm:86]: But it also happens [in this opensuse test](http://angmar.suse.de/tests/2632#step/reboot_and_install/7) for instance

Comment (DimStar77) [reply lib/power_action_utils.pm:86]: I see; that test also just boots from the disk image and never setups the serial console  permissions (conisoletest_setup); which leads to this situation. Which is why http://angmar.suse.de/tests/2632/modules/ssh_key_check/steps/1/src also needs the become_root step in there.

Comment (ggkioulis) [reply lib/power_action_utils.pm:86]: I understand, all I mean is that I should not remove the become_root line, given that it will fail in that specific scenario.

Comment (foursixnine) [reply lib/power_action_utils.pm:86]: still same thing, you need `consoletest_setup` right after `boot_to_desktop`

Comment (ggkioulis) [reply lib/power_action_utils.pm:86]: I 'm all for removing the `become_root` line.
It is just that we should add `consoletest_setup` after `boot_to_desktop` to [gnome+do_not_import_ssh_keys](https://openqa.opensuse.org/tests/1480600#step/reboot_and_install/1), [external_iso]
(https://openqa.opensuse.org/tests/1480703#step/reboot_and_install/1) and [gnome+import_ssh_keys]
(https://openqa.opensuse.org/tests/1480694#step/reboot_and_install/1) so that they don't fail, unless I am misunderstanding something

File: lib/power_action_utils.pm
Comment (okurz) [inline lib/power_action_utils.pm:85]: `x11_start_program('xterm')` should be enough. There should really be no need to make it this complicated. I have seen similar lines in test code which at best is exchanged as well. The line `x11_start_program('xterm')` is much more common

File: lib/power_action_utils.pm
Comment (okurz) [inline lib/power_action_utils.pm:88]: `send_key `alt-f4';` should be enough

File: lib/power_action_utils.pm
Comment (okurz) [inline lib/power_action_utils.pm:91]: I advise against doing something as low level as checking the package version. It should be enough to just turn around the logic check and do something like:

```perl
if (is_sle('<=15-sp2') || is_leap('<=15.2')) {
```

and not do the whole part about reading out RPM versions. No user would do that and the test should reflect clear user scenarios.

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/power_action_utils.pm
Comment (dzedro) [reply lib/power_action_utils.pm:91]: Nonsense, You can't assure what gnome version will be on what version. What use would do has nothing to do with this issue that since gnome version 'x' reboot behavior changed.

File: lib/power_action_utils.pm
Comment (foursixnine) [reply lib/power_action_utils.pm:85]: +1

File: lib/power_action_utils.pm
Comment (foursixnine) [reply lib/power_action_utils.pm:88]: SInce we're in the console already, killall doesn't hurt

File: lib/power_action_utils.pm
Comment (foursixnine) [reply lib/power_action_utils.pm:86]: yep, that's how it should be

File: lib/power_action_utils.pm
Comment (foursixnine) [reply lib/power_action_utils.pm:91]: +1, in this case... gnome package doesn't change the behavior often enough between SLE/Leap versions, to need the change you're proposing, at least in the short future.

File: lib/power_action_utils.pm
Comment (okurz) [reply lib/power_action_utils.pm:88]: it's many more key presses and an unusual way to end a single xterm and not what we do in other cases as well. Tests should be written in a clean and simple way. I know that `killall xterm` is used in other cases where it might make sense, e.g. in tests/x11/sshxterm where one wants to kill all instances of xterm not knowing which window is focussed or where an "alt-f4" is ambiguous. I assume that @ggkioulis just took an existing module as inspiration, which is a good approach, but I am just saying: Other modules are a better source for inspiration here :)

Comment (foursixnine) [reply lib/power_action_utils.pm:88]: Well in this case, and this context: The system is about to be shut down :) So it doesn't really matter

File: lib/power_action_utils.pm
Comment (okurz) [reply lib/power_action_utils.pm:91]: I don't understand why you describe my comment as "nonsense". And regarding what @foursixnine wrote: What I am proposing is used in many other places, also to distinguish between different gnome versions. I don't see a need to do it more complicated here with more runtime code that is passed to the SUT when the test helper module is not about checking for specific package versions of gnome. Our tests should reflect clear user scenarios and be simple to follow and understand. We need to make something different, this is clear, but reading out an RPM version is IMHO a bad system level test design as we are messing internal details that a user does not need to care about. If you all still agree that this is the way to go, go ahead, I am not insisting.

Comment (SergioAtSUSE) [reply lib/power_action_utils.pm:91]: In general, I agree to follow what @okurz is proposing.

Usually, you have tests according to a specific version, so you don't need to check in the code which version are you using, but the test developer knows which version is to be tested and the test reflect that specific version in the test case. And if there are more versions available in the same product, you would have different test cases for each version.

There is no clean way to solve this kind of issue as long as the test code is trying to match at the same time every released and under development product.

Given what we have and in this case, I think it is worth to make an exception. So, I agree to check if the version is "smaller than".

Comment (DimStar77) [reply lib/power_action_utils.pm:91]: It should actually work to do the 'else' branch on all versions, without even caring for the version.

TW currently already uses this when it needs to reboot a Leap 42.x instance before it can reboot; In one case I needed to 'cheat' with a 'empty click which is not needed' - but essentially, it works.

And rebooting via the menu is imho more realistic to be used by users than ctrl-alt-del ever was.

Comment (dzedro) [reply lib/power_action_utils.pm:91]: Because it's nonsense for Maintenance, for TW it's not a problem as it has latest package versions and there is not TW 12 15 ... with older gnome versions. It's problem for e.g. SLE 12-SP2, SLE 15 etc, some versions can get never gnome version tomorrow, some in one month or never, there is no connetion between openQA VERSION and gnome version.

Comment (okurz) [reply lib/power_action_utils.pm:91]: Any service pack of SLE will not suddenly get a new major version of gnome and even a minor version (not patch) should be done explicitly so even if the test fails then this is a good sign that the behaviour should be checked. If openQA tests are confused think about customers that would be confronted with a new gnome version pulled in by *maintenance patches*

Comment (dzedro) [reply lib/power_action_utils.pm:91]: I'm curious why you think service pack won't get new version of gnome ?
This is not first time that interface or behavior of program or DM changed, I'm surprised we still got customers after so many visual changes between versions. You want to check expected behavior, with this view we should drop automation.

Comment (okurz) [reply lib/power_action_utils.pm:91]: I mean no new major version of gnome as maintenance patches during the life time of a service pack

Comment (dzedro) [reply lib/power_action_utils.pm:91]: I don't know if there is any relation between service pack and gnome major version, and if it has to be major version patch for change like this one, I guess no.
With version of distribution there will be failing tests until the update is released. Also TW has staging where will be different gnome version. For maintenance with older supported SLE versions, while each version can get gnome update at any time, will be check of distribution version problem and breaking review.

File: lib/power_action_utils.pm
Comment (ggkioulis) [reply lib/power_action_utils.pm:86]: So, for the current PR, I will leave `become_root` as it is, so that I don't break upstream jobs.
When `consoletest_setup` starts being scheduled before `reboot_and_install` in those jobs, I will submit a new PR and remove `become_root`.
Is this ok?

**Review by okurz — APPROVED** [review]
I don't know how to remove my "request for changes". I guess I will have to "approve" even if I am not confident with all the changes which I described in comments

**Comment by ggkioulis** [comment]
> I don't know how to remove my "request for changes". I guess I will have to "approve" even if I am not confident with all the changes which I described in comments

@SergioAtSUSE proposed that I implement what @DimStar77 suggested, so to just handle rebooting with clicking on all products, so that only needles and some empty clicks change. I am working at this now, and most probably it will not require branching anywhere.

**Comment by ggkioulis** [comment]
> > I don't know how to remove my "request for changes". I guess I will have to "approve" even if I am not confident with all the changes which I described in comments
> 
> @SergioAtSUSE proposed that I implement what @DimStar77 suggested, so to just handle rebooting with clicking on all products, so that only needles and some empty clicks change. I am working at this now, and most probably it will not require branching anywhere.

After a great deal of reneedling, I can eventually say that the clicking our way to reboot for all products (without any branching) is not robust enough. 

This is mainly because after this gnome update there is a greater number of clicks required to perform a reboot (4 clicks) instead of  the ones required in products without the update (3 clicks).
For the non updated products, this fourth click can be bypassed in a hacky way, if a click anywhere needle is created to just sidestep the `assert_and_click('reboot-click-restart');` step. However this needle is volatile and really depends on what background area is considered in every product as an insignificant click, which leads to failures that, to be honest will be hard to debug, since the code side does not really explain this "click somewhere to by pass an assert_and_click" tactic. 
After fighting with a good number of needles trying to make this run on all maintenance products, I can say that it will probably will be a flimsy solution if done without branching.

File: lib/power_action_utils.pm
Comment (foursixnine) [reply lib/power_action_utils.pm:86]: @ggkioulis Sounds good

**Comment by DimStar77** [comment]
We already have a case where we need to cheat with empty-clicks, as even on openSUSE we happen to boot an old OS to upgrade in the end.

See for example https://openqa.opensuse.org/tests/1478203#step/reboot_and_install/7

So imho, the argument that 'clicking does not work' is debunkt

**Review by SergioAtSUSE — APPROVED** [review]
After discussing with @ggkioulis, I realized that a branching would be needed even with clicks.
As a test maintainer/reviewer myself I find the "blind-click" as a problem for future maintainability.

So, I agree with the current PR.

**Comment by ggkioulis** [comment]
> We already have a case where we need to cheat with empty-clicks, as even on openSUSE we happen to boot an old OS to upgrade in the end.
> 
> See for example https://openqa.opensuse.org/tests/1478203#step/reboot_and_install/7
> 
> That test boots a 42.x (i.e. SLE12 level!) system, needs to reboot the gnome desktop and then upgrades to TW. so we already handle the 'tumbleweed based reboot' in a SLE12/Leap 42.x system
> 
> So imho, the argument that 'clicking does not work' is debunkt

I agree that it works, and for the tumbleweed with the non updated gnome it worked fine, but those needles easily break and it is hard to understand what broke in case someone is trying to fix it. I say easily because if it includes a small amount of info it can be matched with smth else, as has happened in [this](http://angmar.suse.de/tests/2664#step/reboot_gnome/4) case, and if it includes a large amount of info as is done in the case you just mentioned, a font change can break the needle and leave the debugger with questions on why this needle is even required.

**Comment by DimStar77** [comment]
No question about debugging, that can be messy.

So as we generally already check what gnome version we have, we essentially also know how many and which clicks are needed. Might be worthy to use this info instead of doing completely different things/reboot methods.

Either way: I can live with the PR

**Review by foursixnine — APPROVED** [review]

**Comment by foursixnine** [comment]
> I don't know how to remove my "request for changes". I guess I will have to "approve" even if I am not confident with all the changes which I described in comments

There's usually a thingie where you can dismiss a review, not sure if it shows for your own reviews though

**Comment by coolo** [comment]
It breaks Z: https://openqa.suse.de/tests/5087708#step/reboot_gnome/20

**Pull request merged**
